### PR TITLE
New model structure and reorg for two routes

### PR DESCRIPTION
**Pull request by maxmcd**

### CODE REVIEW

File: models/models.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+'use strict';
```
Comment (maxmcd) [inline models/models.js]: required for use of `const` and `let`

File: models/models.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          fulfill()
+        }
+      })
+      .catch(reject)      
+    })
+  }
+}
+
+class Message {
+  static create(messageObject) {
```
Comment (maxmcd) [inline models/models.js]: Was going to pass message attributes as options but it quickly got confusing. We might want to put more thought into a message creation process. This doesn't same a lot of lines of code, but would be easy to add lots of complex validation logic.

File: models/models.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      .insert(messageObject).returning("msgid")
+      .then((messageIds) => {
+        fulfill(messageIds[0])
+      }).catch(reject)
+    })
+  }
+}
+
+class Convo {
+  static closeAll(cmid, clid) {
```
Comment (maxmcd) [inline models/models.js]: Would be nice if this had a better name. It is not closing all.

File: models/models.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        .then(function (convos) {
+          db("convos").whereIn("convid", convos)
+            .update({
+              open: false
+            }).then(function (success) {
+              fulfill(success)
+            })
+            .catch(reject)
+        })
+        .catch(reject)
```
Comment (maxmcd) [inline models/models.js]: Sticking with the routes pattern, we can probably return the future on line 45 and then have a single catch statement.

File: routes/cmview.js [hunk 5]
```diff
@@ -859,86 +860,53 @@ router.post("/:cmid/cls/:clid/convos", function (req, res) {
...
+    var newConvoId;
+
+    Convo.closeAll(cmid, clid)
+    .then(() => {
+      return Convo.create(cmid, clid, subject, true)
+    }).then((convoId) => {
+      newConvoId = convoId
+      return Communication.findById(commid)
+    }).then((communication) => {
+      twClient.sendSms({
```
Comment (maxmcd) [inline routes/cmview.js]: As this twilio logic is repeated elsewhere it should likely be abstracted to a utility function as well.

File: models/models.js (hunk 3 above)
Comment (kuanb) [inline models/models.js]: CloseAllSimilar ?

File: routes/cmview.js (hunk 5 above)
Comment (kuanb) [inline routes/cmview.js]: Can you explain line 874? Looks like you create a new func which just returns convo.create()

I assume it is this through this methodology that the `.catch()` can be passed all the way through... 

it is "strange" through . perhaps just me getting used to it. assume you know what you're doing and this does not lead to any risk, just makes the returned function exist "unnested"

**Comment by kuanb** [comment]
looking good. like with prev, i want to do this post implementation of notifications... which i am like 50 % done with. i will focus on it over next 2 days then merge everything at that time... does this inhibit your workflow? if so happy to reshuffle order of things

File: routes/cmview.js (hunk 5 above)
Comment (maxmcd) [inline routes/cmview.js]: Yeah, it's a little weird. Exactly as you're guessing. We know `Convo.create(cmid, clid, subject, true)` returns a future. So we can just return it from this function and the next `then` call will work. 

If the first future errors out, I'm assuming the `.catch` will propogate through all the extra `.then` calls. The login being that a `.then` call on an errored out future should not call the contained function and just return the same future until a `.catch` call is made. So if `closeAll` errors out, we're just running:
`Convo.closeAll(cmid, clid).then().then().then().catch(errorHandler)`

**Comment by maxmcd** [comment]
I'm going to merge this into my branch as-is, will merge with master once your changes get merged in. So no, should be all good.

File: routes/cmview.js (hunk 5 above)
Comment (maxmcd) [inline routes/cmview.js]: As we're now avoiding nested functions, this pattern might also be useful to look at:
https://github.com/caolan/async#waterfalltasks-callback

Gets passed an array of functions and supports error handling. Wildly popular apparently. Avoids nesting and futures. Worth a look.

**Comment by kuanb** [comment]
sorry for delay in pulling this in... naturally it does not conflict with expansions i have been building out... but what needs to happen on my end is that i begin adopting the class system over just rewriting queries all the time, etc.

**Pull request merged**